### ISSUE TITLE
feat(llmjpvl): port LLM-jp-4 VL and Jagle-VL (SigLIP2 tiles + pixel-shuffle mlp1 + Llama/Qwen3 decoder)

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1362]
## Summary

`model_type: "llmjpvl"` covers two Japanese VLM releases that share one architecture: `llm-jp/llm-jp-4-vl-9B-beta` (Llama-style 8B decoder, 196608-token vocab, Harmony-style chat format) and `llm-jp/Jagle-VL-2.2B-Jagle-FineVision` (Qwen3-1.7B decoder with per-head q/k RMSNorm). Both put a SigLIP2-so400m tower (512px, patch 16, 1024 patches per tile) behind dynamic tiling, fold 2x2 patch groups with `pixel_shuffle(0.5)`, and project through a LayerNorm / Linear / GELU / Linear `mlp1` into the decoder width. mlxcel has no `llmjpvl` arm. The port reuses the InternVL runtime shape almost unchanged; the new pieces are the SigLIP tower in place of InternViT, the `llm_config` key, the `<|image_start|>` / `<|image_pad|>` / `<|image_end|>` framing, and a per-backbone text model choice.

## Current behavior

- `src/models/detection.rs` has no `llmjpvl` arm; `get_model_type` fails with `Unsupported model type: llmjpvl`.
- The nearest runtime is InternVL: `src/vision/internvl.rs:143-170` (`InternVLChatVLM { text_model: Qwen2Model, vision_model: InternVitVisionModel, connector: InternVLConnector, processor: InternVLProcessor, image_context_token_id, img_start_token_id, img_end_token_id, num_image_token, eos_token_ids }`), its connector `InternVLConnector::from_weights(weights, "mlp1", eps, downsample_ratio, gs, bits)` at `src/vision/internvl.rs:70-100` (already reads `mlp1.0` LayerNorm, `mlp1.1` and `mlp1.3` Linear and applies `pixel_shuffle`), the tiling processor `src/vision/processors/internvl.rs:45`, and the loader `src/loading/vlm_internvl.rs:98-235`.
- That loader hard-binds the text model to `models::llama3::ModelArgs` parsed from `text_config` (`src/loading/vlm_internvl.rs:110-125`) and reads the tower as `InternVitVisionModel::from_weights(&weights, &vision_config, "vision_model")` (line 158). Neither matches this family: the config nests the decoder under `llm_config`, the tower is SigLIP under `vision_backbone.vision_model.*`, and Jagle-VL needs the Qwen3 decoder (`src/models/qwen3.rs`, which has `forward_with_embeddings` at line 854 and `embed_tokens` at line 864).
- `src/vision/encoders/siglip.rs:521` (`SigLipVisionModel::from_weights_with_quant_and_gelu`) already builds the exact tower (27 layers, 1152 wide, 16 heads, `gelu_pytorch_tanh`).

## Expected behavior

`mlxcel generate -m models/llm-jp-4-vl-9B-beta --image doc.png -p "この画像について説明してください。" -n 128` tiles the image, encodes every tile with SigLIP2, pixel-shuffles `[tiles, 1024, 1152]` to `[tiles, 256, 4608]`, projects to `[tiles, 256, hidden]`, scatters the features onto the `<|image_pad|>` positions, and decodes with the Llama (9B) or Qwen3 (Jagle) backbone.

### config.json

Top level: `model_type: "llmjpvl"`, `architectures: ["LLMjpVLModel"]`, `llm_config` (the decoder config; there is no `text_config`), `vision_config`, `downsample_ratio: 0.5`, `force_image_size: 512`, `img_context_token_id` (14 for llm-jp-4-vl, 151655 for Jagle-VL), `max_dynamic_patch: 12`, `min_dynamic_patch: 1`, `use_thumbnail: true`, `select_layer: -1` (only -1 supported), `ps_version: "v2"`, `template: "llmjp4_harmony"`, `tie_word_embeddings` (false for 9B, true for Jagle-VL).

`llm_config` llm-jp-4-vl-9B: `model_type: "llama"`, `hidden_size 4096`, `num_hidden_layers 32`, `num_attention_heads 32`, `num_key_value_heads 8`, `head_dim 128`, `intermediate_size 14336`, `rope_theta 500000`, `rope_scaling null`, `rms_norm_eps 1e-6`, `vocab_size 196608`, `attention_bias false`, `max_position_embeddings 65536`.

`llm_config` Jagle-VL-2.2B: `model_type: "qwen3"`, `hidden_size 2048`, `num_hidden_layers 28`, `num_attention_heads 16`, `num_key_value_heads 8`, `head_dim 128`, `intermediate_size 6144`, `rope_theta 1000000`, `rms_norm_eps 1e-6`, `vocab_size 151676`, `tie_word_embeddings true`, `eos_token_id 151645`.

`vision_config` (both): `model_type: "siglip_vision_model"`, `hidden_size 1152`, `num_hidden_layers 27`, `intermediate_size 4304`, `num_attention_heads 16`, `patch_size 16`, `image_size 512`, `num_channels 3`, `layer_norm_eps 1e-6`, `hidden_act "gelu_pytorch_tanh"`.

`preprocessor_config.json`: `size {512,512}`, `image_mean/std [0.5,0.5,0.5]`, `resample 2`. `processor_config.json`: `image_seq_length 256`, `max_dynamic_patch 12`, `min_dynamic_patch 1`, `use_thumbnail true`.

### Weights

```
vision_backbone.vision_model.embeddings.patch_embedding.{weight,bias}      # weight [1152,3,16,16] HF layout or [1152,16,16,3] converted
vision_backbone.vision_model.embeddings.position_embedding.weight         # [1024,1152]
vision_backbone.vision_model.encoder.layers.{0..26}.{layer_norm1,layer_norm2}.{weight,bias}
vision_backbone.vision_model.encoder.layers.N.self_attn.{q,k,v,out}_proj.{weight,bias}
vision_backbone.vision_model.encoder.layers.N.mlp.{fc1,fc2}.{weight,bias}
vision_backbone.vision_model.post_layernorm.{weight,bias}
vision_backbone.vision_model.head.*                                       # attention pooling head, skip
mlp1.0.{weight,bias}   # LayerNorm(4608), eps 1e-5
mlp1.1.{weight,bias}   # Linear 4608 -> hidden
mlp1.3.{weight,bias}   # Linear hidden -> hidden
language_model.model.embed_tokens.weight, language_model.model.layers.N.*, language_model.model.norm.weight
language_model.lm_head.weight                                             # absent when tie_word_embeddings
```

### Connector

`x: [tiles, 1024, 1152]` -> reshape `[tiles, 32, 32, 1152]` -> `pixel_shuffle(0.5)` -> `[tiles, 256, 4608]` -> `LayerNorm(eps 1e-5)` -> `Linear(4608, hidden)` -> exact GELU -> `Linear(hidden, hidden)`. This is byte-for-byte the `InternVLConnector::forward` path; the only difference from the InternVL call site is the LayerNorm eps (1e-5, the torch default, not `vision_config.layer_norm_eps`).

### Processor and prompt

Tiling is the InternVL rule (`find_closest_aspect_ratio` by absolute ratio difference, `min_num 1`, `max_num`, `image_size 512`, thumbnail appended when more than one tile), with one extra budget step: `max_num = clamp(((model_max_length - text_tokens) // num_images - 2) // 256 - 1, 1, 12)` where `model_max_length = 4096` (tokenizer config) and `text_tokens` is the token count of the prompt with `<image>` removed. For prompts under roughly 500 tokens this is 12.

Normalization: `(pixel/255 - 0.5) / 0.5`, tiles `[N, 3, 512, 512]`, bicubic resize before cropping.

Each `<image>` placeholder in the user text is replaced by `<|image_start|>` + `<|image_pad|>` x (256 * tiles_i) + `<|image_end|>`; when the text has no placeholder the block is prepended. Token ids: llm-jp-4-vl `<|image_pad|>` 14, `<|image_start|>` 15, `<|image_end|>` 16; Jagle-VL 151655 / 151669 / 151670 (resolve from `added_tokens_decoder`, default to the 9B ids).

Chat template (both checkpoints ship `chat_template.jinja`): `<|start|>system<|message|>{system}<|end|><|start|>user<|message|>{content}<|end|><|start|>assistant`, default system message `You are LLM-jp-VL, a Multimodal LLM trained by LLM-jp.`; the generation prompt must then be extended with `<|channel|>final<|message|>` before tokenization (the processor appends it after the template, so a bare template render is incomplete). No BOS is added (`add_bos_token: false`). Stop ids: 9B `<|return|>` 2 and `<|end|>` 11; Jagle `<|im_end|>` 151645 and `<|return|>` 151675 / `<|end|>` 151672.

## Implementation plan

1. **Detection.** `src/models/detection.rs`: `"llmjpvl" => Ok(ModelType::LlmJpVLM)`. `src/models/mod.rs`: enum variant, supported list, description `("LLM-jp VL (SigLIP2 + pixel-shuffle mlp1 + Llama/Qwen3)", "Other VLM")`, list grouping. Test in `src/models/detection_tests.rs`.

2. **Text backbone abstraction (`src/vision/llmjp_vl.rs`, new).** `enum LlmJpTextModel { Llama(models::llama3::Llama3Model), Qwen3(models::qwen3::Qwen3Model) }` implementing `LanguageModel` by delegation (both inner types already implement `embed_tokens` and `forward_with_embeddings`; Llama at `src/models/llama3.rs:1143-1156`, Qwen3 at `src/models/qwen3.rs:854-866`). Select by `llm_config.model_type`: `"llama"` and `"qwen2"` -> `Llama3Model` (`attention_bias` from config), `"qwen3"` -> `Qwen3Model`; anything else is a load error naming the value.

3. **Runtime (`src/vision/llmjp_vl.rs`).** `LlmJpVlModel { text_model: LlmJpTextModel, vision_model: SigLipVisionModel, connector: InternVLConnector, processor: InternVLProcessor, image_context_token_id, img_start_token_id, img_end_token_id, num_image_token: 256, eos_token_ids, image_seq_length: 256, model_max_length: 4096 }`. `get_input_embeddings(input_ids, pixel_values)`: NCHW -> NHWC, cast to the tower dtype, `vision_model.forward` (post-layernorm output, no CLS), `connector.forward`, then `merge::merge_llava(image_context_token_id, ...)`. Mirror `InternVLChatVLM::get_input_embeddings` (`src/vision/internvl.rs:172-205`).

4. **Loader (`src/loading/vlm_llmjp_vl.rs`, new).** Read the sanitized config; take `llm_config` (fall back to `text_config` if a converted checkpoint renamed it); inherit the top-level `quantization` into it; parse into `llama3::ModelArgs` or `qwen3::ModelArgs` by `model_type`; strip the `language_model.` prefix with `strip_language_model_prefix` (`src/loading/vlm.rs:401`) for the text weights only; build the tower with `SigLipVisionModel::from_weights_with_quant_and_gelu(&weights, &vision_cfg, "vision_backbone.vision_model", gs, bits, false)` after dropping `vision_backbone.vision_model.head.*`; transpose `patch_embedding.weight` to `[O, kH, kW, I]` only when `shape[1] == 3`; build `InternVLConnector::from_weights(&weights, "mlp1", 1e-5, downsample_ratio, gs, bits)`; build `InternVLProcessor::new(force_image_size, min_dynamic_patch, max_dynamic_patch, use_thumbnail)` and set `mean = std = [0.5; 3]` (the struct fields are public, `src/vision/processors/internvl.rs:50-51`). Resolve the three image ids and the stop ids from `tokenizer_config.json` `added_tokens_decoder` by content (`<|image_pad|>`, `<|image_start|>`, `<|image_end|>`, `<|return|>`, `<|end|>`, `<|im_end|>`), defaulting to 14/15/16 and `[2, 11]`. Register in `src/loading/vlm.rs` and dispatch from `src/loading/mod.rs`.

5. **Prompt preparation (`src/multimodal/vlm_runtime.rs`).** Add `VlmRuntimeRef::LlmJpVl` (`src/loaded_model_capabilities.rs`, plus the `LoadedModel::LlmJpVLM` variant and macro arm in `src/loaded_model.rs`). The arm: compute `text_tokens` from the prompt without placeholders, derive `max_num` with the budget formula, call `processor.preprocess_with_tiles` with that `max_dynamic_patch` (add a `preprocess_with_tiles_max(images, max_num)` variant or set the field on a cloned processor), then `insert_internvl_image_tokens(prompt_tokens, &tiles, 256, img_start, img_pad, img_end)` (`src/multimodal/internvl_prompt.rs:58`, which already handles both "bare placeholder present" and "prepend a block" cases), and report `VlmPreparationSummary::LlmJpVl`. Print it in `src/commands/generate_vlm.rs` beside the InternVL arm. Ensure the generation-prompt suffix `<|channel|>final<|message|>` is appended when the rendered template ends with `<|start|>assistant`; the server path renders `chat_template.jinja` through `src/server/chat_template.rs`, so add the suffix rule keyed on `ModelType::LlmJpVLM` where the rendered prompt is finalized.

6. **Output suppression.** `output_suppressed_token_ids` returns the three image ids.

7. **Docs.** `docs/supported-models.md`: add LLM-jp-4 VL and Jagle-VL.

## Validation

(a) Unit tests: `src/vision/llmjp_vl_tests.rs` (`text_model_selection_by_llm_config_model_type`, `connector_uses_torch_default_layer_norm_eps`, `image_budget_formula_caps_at_12_and_floors_at_1`), `src/loading/vlm_llmjp_vl_tests.rs` (`llm_config_is_used_when_text_config_is_absent`, `image_token_ids_resolve_from_added_tokens_decoder`), `src/models/detection_tests.rs`.

(b) Real checkpoints: `llm-jp/Jagle-VL-2.2B-Jagle-FineVision` (bf16 single `model.safetensors`, about 4.4 GB; Qwen3 backbone) and `llm-jp/llm-jp-4-vl-9B-beta` (bf16, 18.1 GB; Llama backbone). Community MLX 4-bit conversions exist (`tokimoa/llm-jp-4-vl-9b-beta-mlx-4bit`); if one fails to load, quantize the original locally.

```
./target/release/mlxcel generate -m models/Jagle-VL-2.2B-Jagle-FineVision --image /tmp/solid_red.png -p "この画像の色は何色ですか。" -n 32
./target/release/mlxcel generate -m models/llm-jp-4-vl-9B-beta --image tests/fixtures/test_image.png -p "この画像について説明してください。" -n 128
```

Acceptance: finite logits; the solid-image answer names red (赤); a 224x224 image reports 1 tile (256 image tokens) and a 1024x768 image reports 12 tiles + thumbnail = 13 blocks of 256; greedy token-exact match against the checkpoints' own `modeling_llmjpvl.py` (CPU bf16) for the first 32 tokens on both checkpoints.

## Acceptance criteria

- [x] `llmjpvl` resolves to `ModelType::LlmJpVLM`; both the Llama and the Qwen3 backbone load from `llm_config`
- [x] SigLIP2 tower loads from `vision_backbone.vision_model.*` with the pooling head skipped and either conv layout accepted
- [x] Image blocks use `<|image_start|>` / `<|image_pad|>` x 256 per tile / `<|image_end|>` with the tile budget formula
- [x] Generation prompt ends with `<|start|>assistant<|channel|>final<|message|>` on both CLI and server paths
- [x] Solid-color image answer is correct on Jagle-VL; first 32 greedy tokens match the checkpoint's own implementation on both models
- [x] docs/supported-models.md updated
- [x] detection table in src/models/detection.rs updated with a test
- [ ] cargo test --workspace --profile test-fast --features metal,accelerate passes (NOT RUNNABLE on the implementation host: Linux aarch64 + CUDA, no Metal and no Accelerate; unrun)
- [x] cargo clippy and cargo fmt pass (run as `cargo clippy --lib --tests --features cuda -- -D warnings` plus `cargo fmt --all`; the `--workspace --all-targets` form is left to the merge gate)

## Out of scope

- `select_layer != -1` (intermediate-layer features): reject at load with a clear error.
- Video input.

---

## Implementation notes (PR #1737)

Validated on Linux aarch64 / NVIDIA GB10 / CUDA sm_121 against both released bf16 checkpoints and their own `modeling_llmjpvl.py` reference (torch 2.14.0+cpu bf16, transformers 4.57.1, greedy). On a 512x512 solid red image mlxcel reproduces the reference exactly on the CLI for both backbones: Jagle-VL 301 prompt tokens and `赤色です。`, llm-jp-4-vl-9B 297 prompt tokens and `この画像は、単色の赤色で塗りつぶされた正方形の図形です。`. Tile counts match the reference too (224x224 gives 1 tile and 256 image tokens, 1024x768 gives 13 tiles and 3328 image tokens). Local checkpoint paths are `models/mlx/jagle-vl-2.2b-jagle-finevision` and `models/mlx/llm-jp-4-vl-9b-beta`, not the `models/<RepoName>` paths this body quotes.

Three things the body did not anticipate. The InternVL "splice after the first token" fallback is wrong for this template, because the first token of a rendered LLM-jp-VL prompt is the system turn's `<|start|>`, so the block is instead spliced after the last `<|start|>user<|message|>` where upstream's `<image>` sits. The server was routing any request with typed media parts to the raw-JSON render, which hands a content list to a template that never inspects typed media; that route is now gated on the template actually reading them. And `InternVLProcessor`'s aspect-ratio tie-break disagreed with the upstream area threshold both families read, returning 1 tile for a 768x768 image at tile size 512 where the reference returns 2x2 plus a thumbnail.

One divergence is recorded rather than fixed: on the server the 9B produces `キャンバス` where the CLI and the reference produce `図形`, stably across three requests. The same server with `--max-batch-size 1` matches the CLI exactly, so it is the batched decode path's reduction order on a near-tie token, not the prompt (which is pinned token-exact against the reference processor by `one_image_prompts_are_token_exact_against_the_reference_processor`).
